### ISSUE TITLE
Check what to do with annotations following acceptation of PEP 649.

### ISSUE THREAD

**Issue by felixchenier** [kineticstoolkit/kineticstoolkit#198]
Currently Kinetics Toolkit uses `from __future__ import annotations` which will become deprecated: https://peps.python.org/pep-0649

Using this was practical because it allowed the implementation of a very simple but okay runtime type checker. I regularly evaluate using beartype, but its output is still too cryptic for non-programmers, and thus could make them panic instead of helping them.

There is no urgency in resolving this, but it still should be fixed.

**Comment by felixchenier** [comment]
Will probably switch to beartype after I'm finished working on https://github.com/beartype/beartype/pull/301

**Comment by felixchenier** [comment]
Related to https://github.com/beartype/beartype/issues/302

My observations on Python typing, type checkers, documentation, to help me take a good direction. Also to explain my thoughts to @leycec (beartype developer) for no real reason.

1. Typing in Python suck. I love Python. But typing really sucks. Big time.
2. Before type hints, we could save our sanity with variable naming such as `str_param1`, `int_param2`. This is what I did in Matlab before switching, because Matlab sucked even more than Python. That sucked because it was both ugly and insecure.
3. With type hints, we now have `param1 : str` and `param2 : int` which sucks a lot less since it is insecure but beautiful. It solves a good part of an important problem which is proper documentation.
4. With type checkers (both static and real-time), we can now theoretically check that we really mean those type hints. BUT (and this is a huge BUT), to please the type checkers, we now have to be less specific in our documentation. Something that was beautiful (`tuple[str, float]`) becomes less specific and uglier (`Sequence[str | float]`) just so that the type checkers do not complain even when highly compatible types (e.g., tuples and lists) are used interchangeably. More secure, but uglier again.
5. And here it is, we just entered the neverending story of asking ourselves: should I please the user or the type checker? I know the type checker will please the user, but at what cost? And then we lose sanity again.

Currently, I am at step "Lost sanity". I'm curious how @leycec didn't lose his before. Or maybe he did. He probably did.

**So maybe a real-time type checker is not what I need after all.**

I have this strategy that I want to evaluate more before jumping the type checking boat. I start from this immutable point: **type hints are not for myself as a code maintainer, but for the user as documentation**.

Therefore:

- I can (and should) use the most precise type hints possible, not in terms of passing a type check, but in terms of the **meaning** of the parameter. For instance, `tuple[float, float]` as more meaning than `Sequence[float]` even if a list of two float would be perfectly valid.
- **After** writing my type hints in a way that helps the user and documentation, then I can factor my whole library so that at least the library respects those type hints, with the help of mypy. Important: I change my code to fit the type hints, not the type hints to fir my code.
- **Then** comes the other problem, what to do if a user of my library runs a function with a totally incompatible argument type. Standard old-school python duck typing would say "Nothing", it will crash at some point anyway. I don't believe this helps the user at all since there will be no way for the user to know what they did wrong. But on the other hand, using a real-type type checker would be way more restrictive, because by design, I decided to use intuitive (but restrictive) type hints for clarity.

Here is what I want to try: **embracing Python's dynamic typing**, while at the same time ensuring that I work with variables of a know type, so that I don't have to put everything in try statements, everytime, and ask myself "but what could happen if...".

The only way I see it possible is not to typecheck, but to typecast. If a variable **can** be converted to a type that I want, then from a Python philosophy, this is perfectly fine. Therefore, if by entering a function, a convert every argument to the type I want, 1) failures will be clear, and 2) I may never trip on types that are defined incompatible but that really are (ex., numpy.Int64 vs int). It will probably be much slower than just type checking, but slower is better than more error-prone. And if things get too slow, I can still optimize the functions that run more often later, to manually type check instead of type casting.

This is not final, it's just a though, but I seriously consider it.

**Comment by felixchenier** [comment]
Unfortunately type casting will also cause problems. Should a function accept a float as an index? If it only uses type casting for ensuring to receive an integer, then it will gladly work with a float. But should it? I guess no.

**Comment by felixchenier** [comment]
Dynamic type checking seems to become more and more broken. I stop wasting energy trying to understand a continuously moving target. I'll do the boring thing and implement implicit boilerplate code in each function, typecasting when appropriate and typechecking when appropriate, on a case-by-case basis.

**Comment by felixchenier** [comment]
OK I jumped out of the rabbit hole. My use of runtime type checking is **only** to help the API user in case of bad parameters. I have more latitude in doing my own tests as boilerplate code instead of keeping up to date with type hints. I still continue testing my type hints statically using MyPy, but I won't reevaluate runtime checking for a good two years, to let the time for python developers to settle on that once for all.

**Comment by leycec** [comment]
**Great sadness.** @beartype fully supports all of the problematic PEPs that everybody else is having trouble with – including PEPs 563, 649, and 695. And you'd worked so hard on those @beartype PRs, too! And they still work great. Admittedly, I haven't documented anything. But I was just about to. I swear.

I understand and appreciate your position, of course. Discussions like beartype/beartype#335 are horrifying and frightening... *but they're only horrifying and frightening for everybody else.* I just opened up that discussion as a Nice Guy™ way of informing the runtime community that CPython 3.13 is about to break them over its knee. But @beartype is perfectly fine with everything.

Come baaaaaaaaaaaack, @felixchenier! Oh, well. Too late, huh? We scared you away and now you've washed your blood-soaked hands of it all. Thankfully, you're still amazing. Let me know if you need anything when you revisit this in a few years. Until then, thanks again for all the great runtime QA you've given us. :pray:

**Comment by felixchenier** [comment]
@leycec Oh, so you read it before I can tell you... I wanted to spare you a bit, but here we are, so I'll tell you the whole story. And just before I begin, it's really an "it's me, not you" situation 😉

OK this will be a long post but you deserve it.

In fact, https://github.com/beartype/beartype/discussions/335 is the final nail of a series of unfortunate events that made me back up, at least for now - and when I speak of years, I usually speak of months.

OK. First, let's hurt more. I was SO CLOSE to release my toolkit with Beartype(TM) TypeChecking Technology. Beartype was fully integrated into the main branch, the dependencies had been updated so that CI works, everything was working perfectly, and the latest features we developed rocked as hell. BUT.

As I continued developing my toolkit, I stumbled on that kind of stupid problem that you could not see coming. I have an issue with... decorators. What??

It's time for a bit of history. 

**Act 1:** In the Eden days of my toolkit, there were no type hints. It was in Python 3.7 era, type hints were something new and I personally found it ugly. "We already write type hints in the docstrings, why would I duplicate it in a function signature that now look so dull, long, strange and ugly?". But then I heard about MyPy and was convinced: Oh, this will help me found those situations where I returned None instead of a string, etc. And so I adopted type hints and MyPy, and I was happy because my code was better.

**Act 2:** My toolkit users sometimes had problems of that kind: `add_event('start', 1.2)` where the function signature was `add_event(time: float, name: str)`. I found that since I already had type hints in all functions, it would be great to find an automated way to gently tell them that we expected a float instead of string, etc., instead of an arcane error message. I told to myself, most probably somebody already did it, there should be some decorator somewhere that does this job. And I found two: typeguard and beartype. I think you may have heard of it. But I had issues with both. Typeguard would not be happy with one of my custom class that uses `from __future__ import annotations`, and beartype would spit incomprehensible jargon instead of "Hey, I need a float, you gave me a string!". I was not happy.

**Act 3:** So I decided to program my own little decorator. I found out that if I use `from __future__ import annotations` everywhere, then I could access `__annotations__` and parse those strings. It was a much easier job than what you did because I know exactly which types I would need to check. It was mainly a sequence of if/elif/else. It worked well, but then I found something disturbing. In case of error, half of the stack trace was calls to decorator wrappers. What a mess in debug time, and my first aim was in fact to clear up the stack trace for the user benefit! I was not happy.

**Act 4:** Therefore I decided to not use a decorator, and to instead do the type checking inside the function, as a first instruction. I think what I did was quite clever, actually. I made a function named `check_types` that would have two arguments: the function itself and the local variables. We would call it that way:

```
# Instead of:

@check_types
def solve_world_problems(world: int, problems: str):
    pass

# I would do:

def solve_world_problems(world: int, problems: str):
    check_types(solve_world_problems, locals())
    pass
```

This way, the `check_type` would have access to `__annotations__` via the function object, and to the parameters via `locals()`. It worked great! I was happy.

**Act 5:** I learned that `from __future__ import annotations` would not stay. The future would not be that future. And if I could not ensure `__annotations__`  to be strings, my little lovely hacky function would stop working. So I remembered that although beartype used a decorator, @leycec managed to NOT pollute the stack trace using clever ways that I couldn't figure out myself, even by inspecting beartype's lovely (eh-em) code. Maybe I could play more with beartype and maybe I could even ask some new features to @leycec, particularly about the arcane exception messages. It turned out that @leycec was a nice guy willing to do lot of work for the others, and in fact, contributing to these additions was actually fun. At the end (which is some weeks ago), my code base was entirely beartype-enabled and I was happy.

**Act 6 - The moment of truth:** In a last round of development for the next release, things began to be more complicated. It appears that I develop a lot using the IPython REPL, and I depend on the %autoreload module magick. It's really painful for me to start a new console every time I make a modification to my modules. Autoreload works very well, UNLESS we mess with it. And decorating functions does mess with it. I get weird errors such as "My function expected a beartype conf, not its parameters". At this moment, my brain was screaming something like "Holy crap, all this for that??". OK, now I must revert everything just because of those decorators?? OK, I did something not that bad previously, with my check_types function. Maybe I could leverage the beartype.DOOR API to do this trick? I tried, but the DOOR API seems to not like `from __future__ import annotations`, at least is has problems with stringified builtins such as "int" or "str". And even if it did, `from __future__ import annotations` will disappear at some time, and I don't know what I'll do. And then, at this very moment, I see your email on the improbability of having sane annotations in the near future. At this moment, I was unhappy because I needed to start thinking everything again.

---

OK, now where's here in the present, with the impression of coming out from a long fever dream. The ONLY thing I wanted from runtime annotations was:

(copy/pasted from above:)

> My toolkit users sometimes had problems of that kind: `add_event('start', 1.2)` where the function signature was `add_event(time: float, name: str)`. I found that since I already had type hints in all functions, it would be great to find an automated way to gently tell them that we expected a float instead of string, etc., instead of an arcane error message.

Wow. What a deep hole I've been in for such a simple problem.

And then, I started to think more and more that I may even have misidentified my problem. Do I really want to adopt strict typing at all? This is not what my need above expresses.

For instance, take this function:

```
def set_color(color: str | tuple[float, float, float]):
    pass
```

Pretty self-descriptive isn't? Even if we don't read the doc, we assume that a color can be defined by a name, or by an RGB tuple. But it would be a shame to get an error if we call it by

```
set_color([1.0, 0.0, 0.0])
```

which is a list and not a tuple. Therefore, if I really want to use runtime type checking, I must define my function as:

```
def set_color(color: str | tuple[float, float, float] | ArrayLike):
    pass
```

But isn't it less clear? Now the user expect three different ways to express colors: by name, by RGB tuple, or by "some kind of array", which in reality is **not** a third way but only a generalization of the tuple.

Then, I could do:

```
def set_color(color: str | ArrayLike):
    pass
```

so there are still only two ways to express a color. But a tuple is not registered as an ArrayLike. Therefore I need to define my own ArrayLike as `ArrayLike | tuple`. But then, it's like cheating, the signature doesn't match what is accepted anymore. Plus, `ArrayLike` is less helpful than `tuple[float, float, float]`.

So I really was turning in circles, and finally I realized that it's not even runtime type checking that I need. Sometimes I need to test the type, sometimes I need to convert compatible parameters to the types I need, and often I need to do completely different tests (an array of how many dimensions, with how many data on each, etc.)

My last-minute workaround was to create this function:

```
def check_param(
    name: str,
    value,
    expected_type,
    *,
    contents_type=None,
    key_type=None,
    length: int | None = None,
    ndims: int | None = None,
    shape: tuple | None = None,
):
    """
    Check that a given parameter has the expected type and optional specs.

    Parameters
    ----------
    name
        Name of the parameter. Will be returned in the exception.
    value
        Value of the parameter.
    expected_type
        Expected type of the parameter.
    contents_type
        Optional. For a tuple or list, ensures that every element is of the
        given type. For a dictionary, ensures that every value is of the
        current type. Does not recurse into nested variables.
    key_type
        Optional. Check that every key of a dict is of a given type. Does not
        recurse into nested variables.
    length
        Optional. Check that a tuple or list has a fixed length.
    ndims
        Optional. Check that an array has a given number of dimensions.
    array_shape
        Optional. Check that an array has a given shape. Use -1 for
        dimensions that do not matter. For instance, to check if an array
        as a shape of Nx4x4, we would use `array_shape = (-1, 4, 4)`.
```

and this function may expand with more and more specific checks that are adapted to my needs. It makes a LOT of boiling plate code in functions with many arguments, but at least it's explicit and tailored to my needs.

I'll try this method for some months, but I stay wildly interested in Python annotations, I love reading on it and I may return to it someday. In fact I expect to come back in a twisted unexpected situation, I just really don't know how and when! Until this moment, it really was a great pleasure to work with you on your loved software and I wish you the best.

**Comment by leycec** [comment]
**Awwwwwwwww.** That was *such* an amazing, feel-good writeup. I humbly appreciate the kindness implicit in investing so much time to this. I too have spent many evenings banging out words.

> I made a function named `check_types` that would have two arguments: the function itself and the local variables.

*...heh.* Yeah. That is clever. That's not much more work than manually decorating everything with `@beartype` and annotating everything with type hints, either. That's even reasonably fast! Eventually, the `beartype.claw.beartype_this_package()` import hook will *probably* outpace that sort of manual approach in both performance and maintainability – but we're *probably* a year or two out from that. I'm writing "*probably*" a lot here.

This is why you're rockin' academia at the Université du Québec and I live in a cabin in the woods. :rofl: 

> So I remembered that although `beartype` used a decorator, @leycec managed to NOT pollute the stack trace using clever ways that I couldn't figure out myself...

You're so nice. I'll take all that ego-stroking and then some. Please! Lavish me with praise!

Seriously, though. I actually can't take credit for that one. Most @beartype inventions were just somebody else's ideas that I "borrowed," ran with, and then fobbed off as my own. This is no exception.

Thankfully, I can explain the dark magic happening here. It's actually kinda silly. So, let's present an equally silly example:

```python
from beartype import beartype, BeartypeConf

@beartype(conf=BeartypeConf(is_debug=True))
def i_heart_cats(cat_scratch_fever: str) -> int:
    return cat_scratch_fever
```

Enabling `is_debug=True` forces @beartype to print the contents of the type-checking wrapper functions it dynamically generates. Pretty handy, huh? The above prints a ton of garbage I will now cruelly subject you to:

```python
(line 0001) def i_heart_cats(
(line 0002)     *args,
(line 0003)     __beartype_get_violation=__beartype_get_violation, # is <function get_func_pith_violation at 0x7f8489077d80>
(line 0004)     __beartype_conf=__beartype_conf, # is "BeartypeConf(is_debug=True)"
(line 0005)     __beartype_func=__beartype_func, # is <function i_heart_cats at 0x7f8489458900>
(line 0006)     **kwargs
(line 0007) ):
(line 0008)     # Localize the number of passed positional arguments for efficiency.
(line 0009)     __beartype_args_len = len(args)
(line 0010)     # Localize this positional or keyword parameter if passed *OR* to the
(line 0011)     # sentinel "__beartype_raise_exception" guaranteed to never be passed.
(line 0012)     __beartype_pith_0 = (
(line 0013)         args[0] if __beartype_args_len > 0 else
(line 0014)         kwargs.get('cat_scratch_fever', __beartype_get_violation)
(line 0015)     )
(line 0016) 
(line 0017)     # If this parameter was passed...
(line 0018)     if __beartype_pith_0 is not __beartype_get_violation:
(line 0019)         # Type-check this parameter or return against this type hint.
(line 0020)         if not isinstance(__beartype_pith_0, str):
(line 0021)             __beartype_violation = __beartype_get_violation(
(line 0022)                 func=__beartype_func,
(line 0023)                 conf=__beartype_conf,
(line 0024)                 pith_name='cat_scratch_fever',
(line 0025)                 pith_value=__beartype_pith_0,
(line 0026)             )
(line 0027) 
(line 0028)             raise __beartype_violation
(line 0029)     # Call this function with all passed parameters and localize the value
(line 0030)     # returned from this call.
(line 0031)     __beartype_pith_0 = __beartype_func(*args, **kwargs)
(line 0032) 
(line 0033)     # Noop required to artificially increase indentation level. Note that
(line 0034)     # CPython implicitly optimizes this conditional away. Isn't that nice?
(line 0035)     if True:
(line 0036)         # Type-check this parameter or return against this type hint.
(line 0037)         if not isinstance(__beartype_pith_0, int):
(line 0038)             __beartype_violation = __beartype_get_violation(
(line 0039)                 func=__beartype_func,
(line 0040)                 conf=__beartype_conf,
(line 0041)                 pith_name='return',
(line 0042)                 pith_value=__beartype_pith_0,
(line 0043)             )
(line 0044) 
(line 0045)             raise __beartype_violation
(line 0046)     return __beartype_pith_0
```

Line 45 is the dark magic. Basically, @beartype separates out these two concerns:

* The **creation** of the exception to be raised in a separate `__beartype_get_violation()` function – which is actually just an alias for `beartype._check.error.errorget.get_func_pith_violation()`, not that anyone particularly cares, but maybe they do. <sup>*they don't*</sup>
* The **raising** of that exception in the `raise __beartype_violation` statement.

Crucially, this means that the type-checking wrapper function **directly raises exceptions.** There are *no* intermediary calls in between. Thus, no useless stack frames that everybody hates. Cool, huh? Wish I'd thought of that. It took two years for somebody else to suggest that. That somebody was probably @TeamSpen210. Pretty sure he single-handedly thought up all my cool ideas.

> It appears that I develop a lot using the IPython REPL, and I depend on the `%autoreload` module magick. It's really painful for me to start a new console every time I make a modification to my modules. Autoreload works very well, UNLESS we mess with it. And decorating functions does mess with it. I get weird errors such as "My function expected a beartype conf, not its parameters".

**Gah!** That's... *horrible.* In theory, @beartype supports module reloading. I have a function that says so. In practice, nobody really supports module reloading. That is something Python *never* did well.

Still, I should hack on that more. If you return to runtime type-checking with grateful tears in your lustrous eyes five years from now, please submit an issue about that! I'd love to see exactly what's going on there. But, please: don't worry about that for now. I feel super-bad that you've expended so much scarce time and resources on this already. :disappointed: 

> Sometimes I need to test the type, sometimes I need to convert compatible parameters to the types I need, and often I need to do completely different tests (an array of how many dimensions, with how many data on each, etc.)

**Totally.** @beartype should definitely have you covered for those tests in a few years – but the conversions will *probably* always remain outside the scope of type-checking, which is a hardship. Still, once @beartype grows BigBoyPants™ in 2025, the following should do *almost* everything you want:

```python
from jaxtyping import Float
from numpy import ndarray

def set_color(color: str | tuple[float, float, float] | list[float, ...] | Float[ndarray, '3']):
```

The only thing that fails to type-check is `list` length when a `list` is passed – which can then be solved with a `beartype.vale.Is[...]` validator. It's a *bit* cumbersome, though. @beartype should *probably* just provide a predefined `beartype.vale.IsLen[...]` validator so you don't have to write your own. Sadly, I am too lazy. This will have to do:

```python
from beartype.vale import Is
from jaxtyping import Float
from numpy import ndarray
from typing import Annotated

RgbList = Annotated[list[float, ...], Is[lst: len(lst) == 3]]
'''
Type hint matching an RGB value in `list` format.
'''

def set_color(color: str | tuple[float, float, float] | RgbList | Float[ndarray, '3']):
```

Indeed, the type-checking rabbit hole could swallow your entire career if you let it. There is danger.

Ultimately, I concede that you've made the Right Choice™ – *for now.* <sup>muhaha</sup> Your `check_param()` does, indeed, do a bit more than @beartype itself is capable of. Okay, a *lot* more. Notably, @beartype currently fails to support three kinds of deep type-checking out-of-the-box that you (and probably everybody else) *really* needs:

```
    contents_type
        Optional. For a tuple or list, ensures that every element is of the
        given type. For a dictionary, ensures that every value is of the
        current type. Does not recurse into nested variables.
    key_type
        Optional. Check that every key of a dict is of a given type. Does not
        recurse into nested variables.
    length
        Optional. Check that a tuple or list has a fixed length.
```

Technically, you *can* type-check those constraints manually with `beartype.vale.Is[...]`-style validators like above – but nobody should have to do that. @beartype should automatically type-check those constraints with standard type hints.

@beartype *does* elegantly type-check array constraints via [`jaxtyping`](https://github.com/patrick-kidger/jaxtyping) – which also produces amazing error messages, praise be to @patrick-kidger:

```
    ndims
        Optional. Check that an array has a given number of dimensions.
    array_shape
        Optional. Check that an array has a given shape. Use -1 for
        dimensions that do not matter. For instance, to check if an array
        as a shape of Nx4x4, we would use `array_shape = (-1, 4, 4)`.
```

Still, it's all "too little, too late" at that point. I agree. Thus, I exhale loudly! :face_exhaling: 

I hope you have a wonderful several years, @felixchenier. My wife and I sure did miss winter this year, though. There was no snow in Ontario. Never saw *that* one coming. There's always snow! It's Ontario. What is this, Florida!? It's kinda hard to ski without snow. We're considering fat bikes for next year. Fat bikes: *when you just wanna roll over everything in your way.* :smile:

**Comment by patrick-kidger** [comment]
FWIW, I generally recommend the decorator approach -- it's possible to catch and manipulate the traceback of a Python exception:
```python
def decorator(fn):
    try:
        ...
    except Exception as e:
        # now mess with e.__traceback__
```
In addition, for specifically IPython, then you can define the `__tracebackhide__ = True` local variable to hide it from IPython stack traces.

As reference points you could see [here](https://github.com/patrick-kidger/equinox/blob/1e601672d38d2c4d483535070a3572d8e8508a20/equinox/_jit.py#L125) or something like `jax._src.traceback_util`.

**Comment by felixchenier** [comment]
Thanks @patrick-kidger for your pointers! Very insightful.

Thanks @leycec for all the energy you put in those discussions! And you know what? I went skiing today, I couldn't believe there was still snow in the Laurentians with the fake winter we got, but there was! It was magical. But right now I'm back to grey-grey-grey Montreal...

**Comment by leycec** [comment]
**Oh, oh!** The Laurentians! We heard *such* awesome things about skiing the Laurentians when we hostelled in Gaspésie a few summers back. All the hostel employees doubled as Laurentian ski guides and trail groomers during the winter months. Adventure tourism just never sleeps in that part of the world. I was jelly af. Now, I must be jelly af of you too. So much jelly. So little snow.

"Grey-grey-grey" sounds about right. Montreal also got hit with the grey stick, huh? This was the greyest non-winter we've ever seen – and we lived in the southern tip of the South Island of New Zealand. Nuthin's greyer than New Zealand, mate. New Zealand is a whole new shade of "grey-grey-grey" known as ["She'll be right."](https://en.wikipedia.org/wiki/She%27ll_be_right) :smile: 

Have a great Spring, @felixchenier. May Summer be mild. Oh, who am I kidding? *Summer's gonna be brutal, bro.* My pale skin tan is burning already.
